Pass to remove prepacking ops.

kimishpatel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #34547 Enable threading for XNNPACK ops.
* **#34319 Pass to remove prepacking ops.**
* #34048 JIT pass to insert XNNPACK ops
* #34047 Integrate XNNPACK with custom class for packing weights.

Summary:
Removes prepacking ops and install them as attributes of the top level
module. Needs to run freezing as the first pass.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20290726](https://our.internmc.facebook.com/intern/diff/D20290726)

kimishpatel committed:
Pass to remove prepacking ops.

Summary:
Removes prepacking ops and install them as attributes of the top level
module. Needs to run freezing as the first pass.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kimishpatel committed:
Update on "Pass to remove prepacking ops."


Summary:
Removes prepacking ops and install them as attributes of the top level
module. Needs to run freezing as the first pass.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20290726](https://our.internmc.facebook.com/intern/diff/D20290726)

[ghstack-poisoned]

jerryzh168 commented on torch/csrc/jit/passes/xnnpack_rewrite.cpp:
```diff
@@ -1,17 +1,83 @@
+      n->kind() == Symbol::fromQualString("xnnpack::conv2d_prepack"));
+}
+
+TypePtr getCustomType(Node* n) {
```
We can get this type from the output of Node, right?

kimishpatel replied:
You mean like, `node->outputs()[0]->type()`? Hmm, let me try.

kimishpatel committed:
Update on "Pass to remove prepacking ops."


Summary:
Removes prepacking ops and install them as attributes of the top level
module. Needs to run freezing as the first pass.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20290726](https://our.internmc.facebook.com/intern/diff/D20290726)

[ghstack-poisoned]

kimishpatel committed:
Update on "Pass to remove prepacking ops."


Summary:
Removes prepacking ops and install them as attributes of the top level
module. Needs to run freezing as the first pass.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20290726](https://our.internmc.facebook.com/intern/diff/D20290726)

[ghstack-poisoned]

kimishpatel committed:
Update on "Pass to remove prepacking ops."


Summary:
Removes prepacking ops and install them as attributes of the top level
module. Needs to run freezing as the first pass.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20290726](https://our.internmc.facebook.com/intern/diff/D20290726)

[ghstack-poisoned]

raghuramank100 commented on torch/csrc/jit/passes/xnnpack_rewrite.cpp:
```diff
@@ -1,17 +1,73 @@
+}
+
+// Must run this pass after constant folding.
+void removePrePackingOps_(script::Module& m) {
```
NIT: Call this FoldPrePackingOps_?

kimishpatel committed:
Update on "Pass to remove prepacking ops."


Summary:
Removes prepacking ops and install them as attributes of the top level
module. Needs to run freezing as the first pass.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20290726](https://our.internmc.facebook.com/intern/diff/D20290726)

[ghstack-poisoned]

jerryzh168 commented on torch/csrc/jit/passes/xnnpack_rewrite.cpp:
```diff
@@ -1,17 +1,82 @@
+    Block* b = blocks_to_visit.top();
+    blocks_to_visit.pop();
+    for (Node* n : b->nodes()) {
+      if (nodeMatchesPackingOps(n)) {
```
I think we can make `nodeMatchesPackingOps` an argument

jerryzh168 commented on torch/csrc/jit/passes/xnnpack_rewrite.cpp:
```diff
@@ -1,17 +1,82 @@
+          TORCH_CHECK(outputs.size() == 1, "Prepack ops have single output");
+          auto attr_name = attr_name_base + c10::to_string(uid++);
+          m.register_attribute(attr_name, n->outputs()[0]->type(), outputs[0]);
+          Value* prepack_op_value = n->outputs()[0];
```
nit: n->output(0), similar for other places

jerryzh168 commented on torch/csrc/jit/passes/xnnpack_rewrite.cpp:
```diff
@@ -1,17 +1,82 @@
+  std::stack<Block*> blocks_to_visit;
+  std::unordered_set<Node*> nodes_to_delete;
+  blocks_to_visit.push(graph->block());
+  int64_t uid = 0;
```
nit: move this closer to where it is used?

kimishpatel committed:
Update on "Pass to remove prepacking ops."


Summary:
Removes prepacking ops and install them as attributes of the top level
module. Needs to run freezing as the first pass.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20290726](https://our.internmc.facebook.com/intern/diff/D20290726)

[ghstack-poisoned]

kimishpatel committed:
Update on "Pass to remove prepacking ops."


Summary:
Removes prepacking ops and install them as attributes of the top level
module. Needs to run freezing as the first pass.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20290726](https://our.internmc.facebook.com/intern/diff/D20290726)

[ghstack-poisoned]

kimishpatel committed:
Update on "Pass to remove prepacking ops."


Summary:
Removes prepacking ops and install them as attributes of the top level
module. Needs to run freezing as the first pass.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20290726](https://our.internmc.facebook.com/intern/diff/D20290726)

[ghstack-poisoned]

kimishpatel committed:
Update on "Pass to remove prepacking ops."


Summary:
Removes prepacking ops and install them as attributes of the top level
module. Needs to run freezing as the first pass.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20290726](https://our.internmc.facebook.com/intern/diff/D20290726)

[ghstack-poisoned]

jerryzh168 commented on torch/csrc/jit/passes/prepack_folding.cpp:
```diff
@@ -0,0 +1,67 @@
+
+// Must run this pass after constant folding.
+void FoldPrePackingOps(script::Module& m,
+    const std::unordered_set<std::string>& foldable_prepacking_ops) {
```
can this take a function argument of Node* -> bool which can check if a given node is a packing op?

kimishpatel replied:
Sure, I can do that. I assume you want to pack not only by name, but maybe if certain conditions satisfy.

jerryzh168 replied:
we might just be looking for name as well, it's just taking functions is more general

jerryzh168 commented on torch/csrc/jit/passes/prepack_folding.cpp:
```diff
@@ -0,0 +1,67 @@
+            graph->insertGetAttr(graph->inputs()[0], attr_name)
+                  ->setType(n->output(0)->type());
+          prepack_op_value->replaceAllUsesWith(packed_weight_attr);
+          //n->removeAllInputs(); //Cannot do this for conv because it will remove constant nodes?
```
remove?

jerryzh168 commented on torch/csrc/jit/passes/prepack_folding.cpp:
```diff
@@ -0,0 +1,67 @@
+    Block* b = blocks_to_visit.top();
+    blocks_to_visit.pop();
+    for (Node* n : b->nodes()) {
+      if (foldable_prepacking_ops.count(n->kind().toQualString()) != 0) {
```
this can be `isPackOp(node)` where `isPackOp` function  is the argument of `FoldPrePackingOps`

kimishpatel replied:
Will make that change.

kimishpatel replied:
Made all requested changes.

kimishpatel committed:
Update on "Pass to remove prepacking ops."


Summary:
Removes prepacking ops and install them as attributes of the top level
module. Needs to run freezing as the first pass.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20290726](https://our.internmc.facebook.com/intern/diff/D20290726)

[ghstack-poisoned]

kimishpatel committed:
Update on "Pass to remove prepacking ops."


Summary:
Removes prepacking ops and install them as attributes of the top level
module. Needs to run freezing as the first pass.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20290726](https://our.internmc.facebook.com/intern/diff/D20290726)

[ghstack-poisoned]

jerryzh168 commented on torch/csrc/jit/passes/xnnpack_rewrite.cpp:
```diff
@@ -1,17 +1,32 @@
 #ifdef USE_XNNPACK
 
 namespace {
+std::unordered_set<std::string>* getFoldablePackingOps() {
```
returning an address is pretty wired, can you just return the set?

jerryzh168 replied:
or define this inside the filter function

kimishpatel replied:
Actually I will just revert to old style of checking. I had this thinking more ops will be added for quantization, but that is not needed with filter function.

jerryzh168 commented on torch/csrc/jit/passes/prepack_folding.cpp:
```diff
@@ -0,0 +1,59 @@
+
+// Must run this pass after constant folding.
+void FoldPrePackingOps(script::Module& m,
+    PrePackingOpsFilterFn is_foldable_op) {
```
lint

kimishpatel committed:
Update on "Pass to remove prepacking ops."


Summary:
Removes prepacking ops and install them as attributes of the top level
module. Needs to run freezing as the first pass.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20290726](https://our.internmc.facebook.com/intern/diff/D20290726)

[ghstack-poisoned]

jerryzh168 reviewed (approved):
LGTM, Thanks!

jerryzh168 commented on test/test_xnnpack_integration.py:
```diff
@@ -366,11 +366,15 @@ def forward(self, x):
             input_data = torch.rand(data_shape)
             ref_result = scripted_model(input_data)
             torch._C._jit_pass_insert_xnnpack_ops(scripted_model._c)
+            if (prepack_removal):
```
nit: parentheses are not needed

dreiss commented on torch/csrc/jit/passes/prepack_folding.cpp:
```diff
@@ -0,0 +1,59 @@
+  // Since this pass can be called by quantization or other passes as well,
+  // we need to make sure we generate a unique "packed_weight_\d+"
+  // Thus static uid.
+  static int64_t uid = 0;
```
This makes me nervous.  It's not thread safe, and running the pass multiple times will change the results, which can be confusing.  Is there some way to keep track of this in the module that's being transformed?

kimishpatel replied:
Thanks for the input David. I missed the thread-safety of this. I think we can borrow module name. Note that the intention here is that this pass will be run after freezing, which does inlining as well. So given that module hierarchy flattened by the time we come here, we can assume that only top level module exists and only for that reason this approach works. 
For now this pass is exposed as a JIT pass on python side for testing independently. But we probably want to hide this from JIT python API and wrap it as a part of "XNNPACKify...". What do you think? @dreiss

dreiss replied:
Yeah, I think we'll want to wrap it in a higher-level function.

But I still think that 

```
A = optimize(torch.jit.load(foo))
B = optimize(torch.jit.load(foo))
```

should produce identical A and B.  With the current code, I don't believe that is the case.

kimishpatel replied:
We can discuss this, but the two passes, 1) convert linear/convs with their packed variants and 2) remove prepacking ops, should pass the test above, especially after fixing this PR for the issue you brought up.
Additionally it should also give us the right answer if we do:
```
A = optimize(torch.jit.load(foo))
B = optimize(A)
```
But you might have caught something I missed.

dreiss commented on torch/csrc/jit/passes/prepack_folding.cpp:
```diff
@@ -0,0 +1,59 @@
+  // Thus static uid.
+  static int64_t uid = 0;
+
+  auto method = m.get_method("forward");
```
Why just forward?  Can we do all methods?

kimishpatel replied:
As mentioned in the previous comment, this will be called after inlining so only forward method is what should matter.

dreiss replied:
What about models with multiple entry points?

kimishpatel replied:
If you mean by, executing model by `model.get_method("foo")(inputs)` then yes, this will fail. The reason for not extending this beyond `forward` was mainly because, this pass relies on freezing pass, which freezes only forward method of the module to be frozen. But this can be fixed.

dreiss commented on torch/csrc/jit/passes/prepack_folding.cpp:
```diff
@@ -0,0 +1,59 @@
+  std::stack<Block*> blocks_to_visit;
+  std::unordered_set<Node*> nodes_to_delete;
+  blocks_to_visit.push(graph->block());
+  std::string attr_name_base("packed_weight_");
```
Can we make this something like `_jit_pass_packed_weight_` to reduce the chance of accidental collisions?

dreiss commented on torch/csrc/jit/passes/prepack_folding.cpp:
```diff
@@ -0,0 +1,59 @@
+          auto outputs = optional_outputs.value();
+          TORCH_CHECK(outputs.size() == 1, "Prepack ops have single output");
+          auto attr_name = attr_name_base + c10::to_string(uid++);
+          m.register_attribute(attr_name, n->output(0)->type(), outputs[0]);
```
Can we check that this is not already registered by the user?

kimishpatel replied:
Do you intend to use that check so as not do registration again? Or delete the attribute and register a new one? I think the latter probably would make sense. I will have to check API to see whats possible here.

dreiss replied:
If it's already present, I think we should throw an exception.  I think it's fine to fail the conversion if a user happens to create an attribute called "_jit_pass_packed_weight_1", but I think we should try not to produce incorrect results even if they do.

kimishpatel replied:
I can add an assert that, although I dont see that happening. But well, we dont write bugs in the code knowingly, so let me add an assert.

kimishpatel committed:
Update on "Pass to remove prepacking ops."


Summary:
Removes prepacking ops and install them as attributes of the top level
module. Needs to run freezing as the first pass.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20290726](https://our.internmc.facebook.com/intern/diff/D20290726)

[ghstack-poisoned]

kimishpatel committed:
Update on "Pass to remove prepacking ops."


Summary:
Removes prepacking ops and install them as attributes of the top level
module. Needs to run freezing as the first pass.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20290726](https://our.internmc.facebook.com/intern/diff/D20290726)

[ghstack-poisoned]

kimishpatel committed:
Update on "Pass to remove prepacking ops."


Summary:
Removes prepacking ops and install them as attributes of the top level
module. Needs to run freezing as the first pass.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20290726](https://our.internmc.facebook.com/intern/diff/D20290726)

[ghstack-poisoned]

kimishpatel committed:
Update on "Pass to remove prepacking ops."


Summary:
Removes prepacking ops and install them as attributes of the top level
module. Needs to run freezing as the first pass.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20290726](https://our.internmc.facebook.com/intern/diff/D20290726)

[ghstack-poisoned]

kimishpatel committed:
Update on "Pass to remove prepacking ops."


Summary:
Removes prepacking ops and install them as attributes of the top level
module. Needs to run freezing as the first pass.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20290726](https://our.internmc.facebook.com/intern/diff/D20290726)

[ghstack-poisoned]

kimishpatel commented:
@dreiss, addressed your comments.

dreiss reviewed (approved):
Looks good.  I think the most important change between the last version and this one is that now every failure mode I can think of will lead to a clear exception, rather than silent strange behavior.  So if anything goes wrong, we'll know quickly and be able to react.

jerryzh168 commented on torch/csrc/jit/passes/prepack_folding.cpp:
```diff
@@ -0,0 +1,66 @@
+              " module of type:",
+              m.type()->name()->qualifiedName(),
+              ". Please make sure that",
+              " FoldPrePackingOps is run at the top level module only.");
```
I think you should increment uid until the attribute is not present, see https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/passes/quantization.cpp#L1733

kimishpatel replied:
I am not follow why that should be done here? The code you pointed to seems to add a child module instance to some parent module. And from my understanding of the code you can try to add multiple child module to the same instance of parent module and thus you need to continue to increment.
This pass is little different in that we want to process this module once only and it should be the top level module. Thus if you try to run this pass on the top level module again and for some reason you come across this condition, then something has gone wrong and you should rather break than continue. Correct me if my understanding is wrong.

kimishpatel committed:
Update on "Pass to remove prepacking ops."


Summary:
Removes prepacking ops and install them as attributes of the top level
module. Needs to run freezing as the first pass.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20290726](https://our.internmc.facebook.com/intern/diff/D20290726)

[ghstack-poisoned]

kimishpatel committed:
Update on "Pass to remove prepacking ops."


Summary:
Removes prepacking ops and install them as attributes of the top level
module. Needs to run freezing as the first pass.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20290726](https://our.internmc.facebook.com/intern/diff/D20290726)

[ghstack-poisoned]

kimishpatel committed:
Update on "Pass to remove prepacking ops."


Summary:
Removes prepacking ops and install them as attributes of the top level
module. Needs to run freezing as the first pass.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20290726](https://our.internmc.facebook.com/intern/diff/D20290726)

[ghstack-poisoned]

kimishpatel committed:
Update on "Pass to remove prepacking ops."


Summary:
Removes prepacking ops and install them as attributes of the top level
module. Needs to run freezing as the first pass.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20290726](https://our.internmc.facebook.com/intern/diff/D20290726)

[ghstack-poisoned]
